File: src/main/java/com/github/svegon/capi/event/render/SpyglassOverlayRenderListener.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.gui.DrawContext;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

public interface SpyglassOverlayRenderListener {
    Event<SpyglassOverlayRenderListener> EVENT = EventFactory.createArrayBacked(SpyglassOverlayRenderListener.class,
            (context, scale, callback) -> {}, listeners -> (context, scale, callback) -> {
        for (SpyglassOverlayRenderListener listener : listeners) {
            listener.onSpyglassOverlayRender(context, scale, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void onSpyglassOverlayRender(DrawContext context, float scale, CallbackInfo callback);
}

File: src/client/kotlin/io/github/svegon/mclientapi/client/event/render/SpyglassOverlayRenderListener.kt
package io.github.svegon.mclientapi.client.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.gui.DrawContext
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.util.function.Function

fun interface SpyglassOverlayRenderListener {
    fun onSpyglassOverlayRender(context: DrawContext, scale: Float, callback: CallbackInfo)

    companion object {
        @JvmField
        val EVENT: Event<SpyglassOverlayRenderListener> =
            EventFactory.createArrayBacked(SpyglassOverlayRenderListener::class.java,
                SpyglassOverlayRenderListener { context: DrawContext, scale: Float, callback: CallbackInfo -> }
            ) { listeners: Array<SpyglassOverlayRenderListener> ->
                SpyglassOverlayRenderListener { context: DrawContext, scale: Float, callback: CallbackInfo ->
                    for (listener in listeners) {
                        listener.onSpyglassOverlayRender(context, scale, callback)

                        if (callback.isCancelled) {
                            return@SpyglassOverlayRenderListener
                        }
                    }
                }
            }
    }
}
